<h1>Mencari Handling missing value</h1>

In [93]:
import pandas as pd
df=pd.read_csv('/content/top_10000_1960-now (1).csv')

df.isnull().sum()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 35 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Track URI             9999 non-null   object 
 1   Track Name            9998 non-null   object 
 2   Artist URI(s)         9997 non-null   object 
 3   Artist Name(s)        9998 non-null   object 
 4   Album URI             9997 non-null   object 
 5   Album Name            9998 non-null   object 
 6   Album Artist URI(s)   9997 non-null   object 
 7   Album Artist Name(s)  9997 non-null   object 
 8   Album Release Date    9997 non-null   object 
 9   Album Image URL       9995 non-null   object 
 10  Disc Number           9999 non-null   int64  
 11  Track Number          9999 non-null   int64  
 12  Track Duration (ms)   9999 non-null   int64  
 13  Track Preview URL     7102 non-null   object 
 14  Explicit              9999 non-null   bool   
 15  Popularity           

<h1>Mencari Handling duplikat value</h1>

In [ ]:
df_house = pd.read_csv('/content/top_10000_1960-now (1).csv')

duplicate = df_house.duplicated().sum()
print('Jumlah nilai duplikat pada data: ', duplicate)

Jumlah nilai duplikat pada data:  48


<h1>Mencari Handling outlier (Menggunakan metode IQR)</h1>

In [ ]:
import pandas as pd
import numpy as np


file_path = '/content/top_10000_1960-now (1).csv'
df = pd.read_csv(file_path)

# Identify numeric columns
numeric_cols = df.select_dtypes(include=np.number).columns


Q1 = df[numeric_cols].quantile(0.25)
Q3 = df[numeric_cols].quantile(0.75)


IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Detect outliers in numeric columns
outliers = (df[numeric_cols] < lower_bound) | (df[numeric_cols] > upper_bound)

print("Number of outliers in each column:")
print(outliers.sum())

# Optionally, remove outliers from the DataFrame
df_cleaned = df[~outliers.any(axis=1)]

# Display the cleaned dataset (without outliers)
print("\nCleaned dataset preview:")
print(df_cleaned.head())

Number of outliers in each column:
Disc Number             233
Track Number            329
Track Duration (ms)     284
Popularity                0
Danceability             72
Energy                   51
Key                       0
Loudness                216
Mode                      0
Speechiness            1051
Acousticness            479
Instrumentalness       2065
Liveness                499
Valence                   0
Tempo                   231
Time Signature          470
Album Genres              0
dtype: int64

Cleaned dataset preview:
                              Track URI  \
2  spotify:track:70XtWbcVZcpaOddJftMcVi   
5  spotify:track:4bEb3KE4mSKlTFjtWJQBqO   
7  spotify:track:5LjSxAIKwyZvQqJ04ZQ0Da   
8  spotify:track:00qOE7OjRl0BpYiCiweZB2   
9  spotify:track:3OalxlWH0v14kyBcNBMINt   

                                          Track Name  \
2                 From the Bottom of My Broken Heart   
5                         Don't Stop - 2004 Remaster   
7  Something About The 

<h1>Normalisasi kolom numerik</h1>

In [ ]:
from sklearn.preprocessing import MinMaxScaler
df_iris = pd.read_csv('/content/top_10000_1960-now (1).csv')

#Menginisialisasi Min Max Scaller
scaler = MinMaxScaler(feature_range=(0,1))
#Numerical Columns
num_col_ = [col for col in df_iris.columns if
df_iris[col].dtypes != 'O' and col != 'Id']
x1 = df_iris[num_col_]



x1 = scaler.fit_transform(x1)
x1 = pd.DataFrame(x1, columns=num_col_)
x1.head()

/usr/local/lib/python3.10/dist-packages/sklearn/utils/_array_api.py:695: RuntimeWarning: All-NaN slice encountered
  return xp.asarray(numpy.nanmin(X, axis=axis))
/usr/local/lib/python3.10/dist-packages/sklearn/utils/_array_api.py:712: RuntimeWarning: All-NaN slice encountered
  return xp.asarray(numpy.nanmax(X, axis=axis))


Disc Number  Track Number  Track Duration (ms)  Explicit  Popularity  \
0     0.066667      0.032258             0.138534       0.0    0.000000   
1     0.066667      0.032258             0.151890       0.0    0.653061   
2     0.066667      0.064516             0.200196       0.0    0.571429   
3     0.066667      0.118280             0.149507       0.0    0.428571   
4     0.066667      0.096774             0.287432       0.0    0.000000   

   Danceability    Energy       Key  Loudness  Mode  Speechiness  \
0      0.624494  0.874621  0.727273  0.530946   1.0     0.067511   
1      0.835020  0.745231  0.181818  0.727293   1.0     0.209564   
2      0.685223  0.666994  0.636364  0.752933   1.0     0.042897   
3      0.691296  0.730185  0.818182  0.636276   1.0     0.364276   
4      0.322874  0.628879  0.000000  0.614774   1.0     0.096624   

   Acousticness  Instrumentalness  Liveness   Valence     Tempo  \
0      0.015941          0.113706  0.405322  0.506533  0.511479   
1      0.014326          0.000022  0.230297  0.804020  0.583008   
2      0.565085          0.000001  0.333675  0.709548  0.344087   
3      0.573157          0.000052  0.027021  0.837186  0.345601   
4      0.681129          0.000074  0.283521  0.499497  0.393818   

   Time Signature  Album Genres  
0             0.8           NaN  
1             0.8           NaN  
2             0.8           NaN  
3             0.8           NaN  
4             0.8           NaN

<h1>Encoding kolom kategorikal (Menggunakan one hot encoding)</h1>

In [ ]:
file_path = '/content/top_10000_1960-now (1).csv'
df = pd.read_csv(file_path)


categorical_columns = df.select_dtypes(include=['object']).columns
print("Kolom kategorikal:", categorical_columns)


df_one_hot_encoded = pd.get_dummies(df, columns=categorical_columns)

# Tampilkan hasil One-Hot Encoding
print("\nDataset setelah One-Hot Encoding:")
print(df_one_hot_encoded.head())


Kolom kategorikal: Index(['Track URI', 'Track Name', 'Artist URI(s)', 'Artist Name(s)',
       'Album URI', 'Album Name', 'Album Artist URI(s)',
       'Album Artist Name(s)', 'Album Release Date', 'Album Image URL',
       'Track Preview URL', 'ISRC', 'Added By', 'Added At', 'Artist Genres',
       'Label', 'Copyrights'],
      dtype='object')

Dataset setelah One-Hot Encoding:
   Disc Number  Track Number  Track Duration (ms)  Explicit  Popularity  \
0            1             3               216270     False           0   
1            1             3               237120     False          64   
2            1             6               312533     False          56   
3            1            11               233400     False          42   
4            1             9               448720     False           0   

   Danceability  Energy  Key  Loudness  Mode  ...  \
0         0.617   0.872  8.0   -12.305   1.0  ...   
1         0.825   0.743  2.0    -5.995   1.0  ...   
2       

<h1>Membuat Feature engineering</h1>

In [ ]:
import pandas as pd

# Load the dataset
file_path = '/content/top_10000_1960-now (1).csv'
data = pd.read_csv(file_path)

# 1. Extracting Year from Album Release Date
data['Album Release Date'] = pd.to_datetime(data['Album Release Date'], errors='coerce')
data['Release Year'] = data['Album Release Date'].dt.year

# 2. Create an 'Audio Profile' by averaging audio features
audio_features = ['Speechiness', 'Acousticness', 'Instrumentalness', 'Liveness', 'Valence']
data['Audio Profile'] = data[audio_features].mean(axis=1)

# 3. Word Count for Track Name, Artist Name, and Album Name
data['Track Name Word Count'] = data['Track Name'].apply(lambda x: len(str(x).split()))
data['Artist Name Word Count'] = data['Artist Name(s)'].apply(lambda x: len(str(x).split()))
data['Album Name Word Count'] = data['Album Name'].apply(lambda x: len(str(x).split()))

# 4. Categorize the release year into Decades
data['Decade'] = (data['Release Year'] // 10) * 10

# 5. Length of the Song Title (character count)
data['Track Name Length'] = data['Track Name'].apply(lambda x: len(str(x)))


output_file = '/content/top_10000_1960-now_featured.csv'
data.to_csv(output_file, index=False)


data[['Track Name', 'Release Year', 'Audio Profile', 'Track Name Word Count', 'Decade', 'Track Name Length']].head()

Track Name  Release Year  Audio Profile  \
0  Justified & Ancient - Stand by the Jams        1992.0       0.217560   
1          I Know You Want Me (Calle Ocho)        2009.0       0.240044   
2       From the Bottom of My Broken Heart        1999.0       0.326900   
3         Apeman - 2014 Remastered Version        2014.0       0.339690   
4       You Can't Always Get What You Want        1969.0       0.305955   

   Track Name Word Count  Decade  Track Name Length  
0                      8  1990.0                 39  
1                      7  2000.0                 31  
2                      7  1990.0                 34  
3                      5  2010.0                 32  
4                      7  1960.0                 34

<h1>Splitting data ke dalam data training & data testing</h1>

In [ ]:
from sklearn.model_selection import train_test_split
import pandas as pd

# Load the dataset
df = pd.read_csv('top_10000_1960-now (1).csv')

# Subset feature dan target
X = df.iloc[:, :-1]
y = df.iloc[:, -1]


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


print('Dimensi X train: ', X_train.shape)
print('Dimensi X test: ', X_test.shape)
print('Dimensi y train: ', y_train.shape)
print('Dimensi y test: ', y_test.shape)


Dimensi X train:  (7999, 34)
Dimensi X test:  (2000, 34)
Dimensi y train:  (7999,)
Dimensi y test:  (2000,)


<H1>PENJELASAN</H1>

<H1>1. Mencari Handling Missing Value</H1> Pada bagian ini, dataset *top_10000_1960-now.csv* dimuat dan dilakukan pengecekan terhadap missing values pada setiap kolom. Nilai yang hilang (missing values) perlu diidentifikasi untuk diambil langkah pembersihan data sebelum analisis atau modeling.

    Kode: Dataset dimuat menggunakan pandas, dan jumlah missing values di setiap kolom dihitung dengan metode .isnull().sum().
- Hasil: Beberapa kolom memiliki missing values, di antaranya kolom:

    Track Name: 1 nilai hilang.
    Artist URI(s): 2 nilai hilang.
    Album Image URL: 4 nilai hilang.
    Track Preview URL: 2897 nilai hilang.
    Artist Genres: 550 nilai hilang.
    Album Genres: 9999 nilai hilang.

- Efek: Karena banyak kolom yang memiliki missing values, terutama Track Preview URL, Artist Genres, dan Album Genres, perlu dilakukan penanganan. Ini bisa berupa mengisi missing values dengan nilai rata-rata, median, atau bahkan menghapus kolom dengan terlalu banyak missing values.

<H1>2. Mencari Handling Duplikat Value</H1> Langkah selanjutnya adalah mencari nilai yang duplikat dalam dataset. Duplikasi dapat mempengaruhi hasil analisis atau model, karena data yang sama dihitung lebih dari satu kali.

    Kode: Untuk mengecek apakah ada data yang duplikat, digunakan metode .duplicated().sum() untuk menghitung jumlah baris yang terduplikat.

    Hasil: Belum terlihat hasil spesifik terkait jumlah duplikat pada bagian ini.

    Efek: Jika ditemukan baris yang terduplikasi, baris tersebut perlu dihapus dengan metode .drop_duplicates() untuk memastikan dataset tetap bersih dan analisis lebih akurat.

<H1>Kesimpulan:</H1> - *Handling Missing Values:* Dataset *top_10000_1960-now.csv* memiliki beberapa kolom dengan missing values yang cukup signifikan, terutama di kolom *Track Preview URL* dan *Genres*.

    Handling Duplikat: Langkah pengecekan duplikat perlu dilakukan untuk memastikan dataset bersih dari data yang berulang, namun hasil pada bagian ini masih perlu ditinjau lebih lanjut.

# Deep Learning

In [97]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping

In [98]:
data = pd.read_csv("top_10000_1960-now (1).csv")

In [99]:
features = ['Speechiness', 'Acousticness', 'Instrumentalness', 'Liveness', 'Tempo']
target = 'Valence'

In [100]:
data_cleaned = data[features + [target]].dropna()

In [101]:
data_cleaned['Valence'] = (data_cleaned['Valence'] >= 0.5).astype(int)

In [102]:
X = data_cleaned[features]
y = data_cleaned['Valence']

In [103]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

In [104]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)


In [85]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [105]:
model = Sequential([
    Dense(128, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    Dropout(0.3),
    Dense(64, activation='relu'),
    Dropout(0.3),
    Dense(32, activation='relu'),
    Dropout(0.3),
    Dense(16, activation='relu'),
    Dropout(0.3),
    Dense(1, activation='sigmoid')
])

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [120]:
model.compile(optimizer=tf.keras.optimizers.RMSprop(learning_rate=0.001),
              loss='binary_crossentropy', metrics=['accuracy'])

In [121]:
history = model.fit(
    X_train_scaled, y_train,
    epochs=50,
    batch_size=16,
    validation_data=(X_val_scaled, y_val),
)

Epoch 1/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 12s 11ms/step - accuracy: 0.6752 - loss: 0.6199 - val_accuracy: 0.6790 - val_loss: 0.6055
Epoch 2/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.6700 - loss: 0.6231 - val_accuracy: 0.6755 - val_loss: 0.6074
Epoch 3/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6697 - loss: 0.6280 - val_accuracy: 0.6770 - val_loss: 0.6064
Epoch 4/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6640 - loss: 0.6238 - val_accuracy: 0.6785 - val_loss: 0.6088
Epoch 5/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6753 - loss: 0.6201 - val_accuracy: 0.6810 - val_loss: 0.6092
Epoch 6/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6712 - loss: 0.6184 - val_accuracy: 0.6790 - val_loss: 0.6073
Epoch 7/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6590 - loss: 0.6293 - val_accuracy: 0.6760 - val_loss: 0.6092
Epoch 8/50
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6853 - loss: 0.6149 - val_accuracy: 

In [122]:
train_loss, train_accuracy = model.evaluate(X_train_scaled, y_train, verbose=0)
val_loss, val_accuracy = model.evaluate(X_val_scaled, y_val, verbose=0)


In [123]:
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")
print(f"Validation Accuracy: {val_accuracy * 100:.2f}%")

Training Accuracy: 67.73%
Validation Accuracy: 68.15%
